mongo/integration/sdam_error_handling_test.go

```diff
@@ -78,16 +79,59 @@ func TestSDAMErrorHandling(t *testing.T) {
 				mt.ResetClient(clientOpts)
 				clearPoolChan()
 
-				// The saslContinue blocks for 150ms so run the InsertOne with a 100ms context to cause a network
-				// timeout during auth and assert that the pool was cleared.
+				// Use a context with a 100ms timeout so that the saslContinue delay of 150ms causes
+				// an operation-scoped context timeout (i.e. a timeout not caused by a client timeout
+				// like connectTimeoutMS or socketTimeoutMS).
 				timeoutCtx, cancel := context.WithTimeout(mtest.Background, 100*time.Millisecond)
 				defer cancel()
 				_, err := mt.Coll.InsertOne(timeoutCtx, bson.D{{"test", 1}})
 				assert.NotNil(mt, err, "expected InsertOne error, got nil")
 				assert.True(mt, mongo.IsTimeout(err), "expected timeout error, got %v", err)
 				assert.True(mt, mongo.IsNetworkError(err), "expected network error, got %v", err)
+				assert.False(mt, isPoolCleared(), "expected pool not to be cleared but was cleared")
+			})
+
+			mt.Run("pool cleared on non-operation-scoped network timeout", func(mt *mtest.T) {
+				// Assert that the pool is cleared when a connection created by an application
+				// operation thread encounters a timeout caused by connectTimeoutMS during
+				// handshaking.
+
+				appName := "authConnectTimeoutTest"
+				// Set failpoint on saslContinue instead of saslStart because saslStart isn't done when using
+				// speculative auth.
+				mt.SetFailPoint(mtest.FailPoint{
+					ConfigureFailPoint: "failCommand",
+					Mode: mtest.FailPointMode{
+						Times: 1,
+					},
+					Data: mtest.FailPointData{
+						FailCommands:    []string{"saslContinue"},
+						BlockConnection: true,
+						BlockTimeMS:     150,
+						AppName:         appName,
+					},
+				})
+
+				// Reset the client with the appName specified in the failpoint.
+				clientOpts := options.Client().
+					SetAppName(appName).
+					SetRetryWrites(false).
+					SetPoolMonitor(poolMonitor).
+					// Set a 100ms socket timeout so that the saslContinue delay of 150ms causes a
+					// timeout during socket read (i.e. a timeout not caused by the InsertOne context).
+					SetSocketTimeout(100 * time.Millisecond)
+				mt.ResetClient(clientOpts)
+				clearPoolChan()
+
+				// Use context.Background() so that the new connection will not time out due to an
+				// operation-scoped timeout.
+				_, err := mt.Coll.InsertOne(context.Background(), bson.D{{"test", 1}})
+				assert.NotNil(mt, err, "expected InsertOne error, got nil")
+				assert.True(mt, mongo.IsTimeout(err), "expected timeout error, got %v", err)
+				assert.True(mt, mongo.IsNetworkError(err), "expected network error, got %v", err)
 				assert.True(mt, isPoolCleared(), "expected pool to be cleared but was not")
 			})
+
 			mt.RunOpts("pool cleared on non-timeout network error", noClientOpts, func(mt *mtest.T) {
```

Comment:
This test appeared to timeout in [this task](https://evergreen.mongodb.com/lobster/evergreen/test/mongo_go_driver_tests_42_plus_zlib_zstd_support__version~4.4_os_ssl_40~ubuntu1804_64_go_1_16_test_replicaset_auth_ssl_patch_4adb6d73ca0fe53c99e11765005f37033acd6471_60d56ac30ae60671e32a3517_21_06_25_05_33_57/0/be7221d15d9e87e5280e416c7e66b9d9#bookmarks=0%2C17748&l=1&shareLine=11762). I assume that is timing out on the channel read, indicating a pool clear does not occur.

It is not clear to me why. Given that it is not a consistent failure, one thing that does look suspicious to me is the fact that `clearPoolChan()` is called after `mt.ResetClient`. That looks like a race to me (monitoring starts and immediately clears the pool, then the channel is cleared, then the channel is read from).

Reply:
Ah, good catch on the timeout failure. The test that timed out was actually in the `"after handshake completes"` block, which I've seen fail intermittently before and have so far been ignoring. Will look further into the timeout failure and see if other intermittent test failures in `sdam_error_handling_test.go` are related.

Reply:
I've replaced all uses of the shared `poolMonitor` and associated `clearPoolChan()` and `isPoolCleared()` functions with individual instances of the new `testPoolMonitor` type. The tests seem to reliably pass now.